ulam-vs-random.py

```diff
@@ -1,19 +1,30 @@
 #!/usr/bin/env python3
 import random
 
-
+"""Seives Algo to find prime numbers within n . More Time Efficient. """
 def get_primes(n):
-    """Get all primes less than or equal to n."""
-    if n < 2:
-        return []
-    primes = [2]
-    for i in range(3, n+1):
-        for p in primes:
-            if i % p == 0:
-                break
-            if p > int(i ** 0.5):
-                primes.append(i)
-                break
+
+    primes = []
+    
+    if(n<2):
+        return primes;
+
+
+    prime = [True for i in range(n+1)]
+ 
+    i = 2
+    while (i * i <= n):
+		
+        if (prime[i] == True):
+			
+            for j in range(i * 2, n+1, i):
```

Comment:
The optimal version of Eratosthenes' sieve starts checking `j`s from `i * i`. Any composites less than `i * i` have already been marked as a composite so it is unnecessary to check them again.

Proof: A number x such that x < i * i does not have any prime factors greater than or equal to i. If all its prime factors are less than i, then we have iterated on all multiples of those factors in a previous iteration of the outer loop during which x would have been marked as composite.